File: LinkedLists/206_reverse.py
from parse_input import *


# O(N)
class Solution:
    def reverseList(self, head):
        prev = None
        curr = head

        while curr:
            nxt = curr.next
            curr.next = prev

            prev = curr
            curr = nxt

        return prev


if __name__ == '__main__':
    nums = input('Space separated values of l: ').strip()
    l = str2List(nums)

    l_sum = Solution().reverseList(l)
    print('Reversed: ', List2str(l_sum))

File: LinkedLists/206_reverse.cpp
#include <string>
#include <iostream>
#include <sstream>

#include "parse_input.h"


// O(N)
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;

        while(curr){
            ListNode* next = curr->next;
            curr->next = prev;

            prev = curr;
            curr = next;
        }

        return prev;
    }
};


int main(){
    std::string line;
    std::cout << "Space separated values of l: ";
    std::getline(std::cin, line);
    ListNode* l = str2List(line);

    ListNode* l_res = Solution().reverseList(l);
    std::string answer;
    List2str(l_res, answer);
    std::cout << "Reversed: " << answer << std::endl;

    return 0;
}